### PR TITLE
DOXYGEN/API: Documentation update

### PR DESCRIPTION
**Pull request by shamisp**
* Adding diagram to the HTML version of the document
* Updating documentation for the UCP API

### CODE REVIEW

**Comment by shamisp** [comment]
autobot build

**Comment by shamisp** [comment]
@yosefe - please take a look.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1049/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1050/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/479/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/480/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -601,14 +651,47 @@ void ucp_ep_destroy(ucp_ep_h ep);
...
- * @param [in]     context    UCP context to map memory on.
- * @param [out]    address_p  If != NULL, memory region to map.
- *                            If == NULL, filled with a pointer to allocated region.
- * @param [in]     length     How many bytes to allocate.
+ * @brief Map or allocate memory for zero-copy operations.
+ *
+ * This routine maps or/and allocates a user-specified memory segment with @ref
+ * ucp_context_h "UCP application context" and the network resources associated
+ * with it. If the application specifies NULL as an address for the memory
+ * segment, the routine allocates and maps the memory segment.  The network
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: can say the address is returned in this argument

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -601,14 +651,47 @@ void ucp_ep_destroy(ucp_ep_h ep);
...
+ * stack associated with an application context can typically send and receive
+ * data from the mapped memory without CPU intervention; some devices and
+ * associated network stacks require the memory to be mapped to send and
+ * receive data. The @ref ucp_mem_h "memory handle" includes all information
+ * required to access the memory locally using UCP routines, , while @ref
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: two ,

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -601,14 +651,47 @@ void ucp_ep_destroy(ucp_ep_h ep);
...
+ * ucp_rkey_h "remote registration handle" provides an information that is
+ * necessary for remote memory access.
+ *
+ * @note
+ * Another well know terminology for the "map" operation that is typically
+ * used in the context of networking is memory "registration". The UCP library
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: or "pinning"

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -601,14 +651,47 @@ void ucp_ep_destroy(ucp_ep_h ep);
...
+ * registers the memory the available hardware so it can be assessed directly
+ * by the hardware.
+ *
+ * Memory mapping assumptions:
+ * @li A given memory segment can be mapped by several different communication
+ * stacks, if these are compatible.
+ * @li The pointer returned may be used with any sub-region of the mapped
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: The handle returned

File: src/ucp/api/ucp.h [hunk 5]
```diff
@@ -601,14 +651,47 @@ void ucp_ep_destroy(ucp_ep_h ep);
...
+ * by a user, then the user is responsible for segmentation and subsequent
+ * management.
+
+ * @param [in]     context    Application @ref ucp_context_h "context" to map
+ *                            (register) and allocate the memory on.
+ * @param [out]    address_p  If the pointer to the address is not NULL,
+ *                            the routine maps (registers) the memory segment.
+ *                            if the pointer is NULL, the library allocates
+ *                            mapped (registered) memory segment.
+ * @param [in]     length     Length (in bytes) to allocate.
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: or to register

File: src/ucp/api/ucp.h [hunk 6]
```diff
@@ -616,23 +699,62 @@ ucs_status_t ucp_mem_map(ucp_context_h context, void **address_p, size_t length,
...
 /**
  * @ingroup UCP_MEM
- * @brief Undo the operation of uct_mem_map().
+ * @brief Unmap memory segment
+ *
+ * This routine unmaps a user specified memory segment, that was previously
+ * mapped using the @ref ucp_mem_map "ucp_mem_map()" routine.  The unmap
+ * routine will also release the resources associated with the memory
+ * @ucp_mem_h "handle".  When the function returns, the @ref ucp_mem_h will be
+ * invalid and cannot be used with any UCP routine.
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: and the associated remote key also becomes invalid for remote access

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1053/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1054/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/482/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/483/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1055/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/484/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1056/ for details.

**Comment by shamisp** [comment]
@yosefe please take a look.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1057/ for details.

**Comment by yosefe** [comment]
:+1:

**Pull request merged**